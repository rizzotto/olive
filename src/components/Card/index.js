import {
  Card as MaterialCard,
  CardActionArea,
  CardActions,
  CardContent,
  CardMedia,
  CircularProgress,
  Collapse,
  IconButton,
  Grow,
  Typography,
} from '@material-ui/core';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import FavoriteIcon from '@material-ui/icons/Favorite';
import clsx from 'clsx';
import PropTypes from 'prop-types';
import React, { useContext } from 'react';
import { ReactComponent as Clock } from '../../assets/clock.svg';
import { ReactComponent as Plate } from '../../assets/plate.svg';
import getPlaceholder from '../../constants/placeholderIcons';

import { AppContext } from '../../context/AppProvider';
import RecipesAPI from '../../services/RecipesAPI';
import useStyles from './styles';

function Card({ data, onClickCard, showStatus, ...others }) {
  const {
    description,
    difficulty,
    logo,
    portions,
    preparation_time,
    recipe_id,
    status,
    title,
  } = data;

  const {
    updateFavoriteRecipes,
    favouriteRecipes,
    setFavouriteRecipes,
  } = useContext(AppContext);

  const classes = useStyles({ difficulty, status, showStatus });
  const [expanded, setExpanded] = React.useState(false);
  const [isLiked, setIsLiked] = React.useState(
    favouriteRecipes && favouriteRecipes.includes(data.recipe_id),
  );

  const handleExpandClick = () => {
    setExpanded(!expanded);
  };

  const handleHeartClick = async () => {
    const fav = await updateFavoriteRecipes(data.recipe_id);

    if (!isLiked) RecipesAPI().likeRecipe(data.recipe_id);
    else RecipesAPI().dislikeRecipe(data.recipe_id);

    setFavouriteRecipes(fav);
    setIsLiked(!isLiked);
  };

  const handleOnClickCard = () => {
    if (onClickCard) onClickCard(data);
  };

  const difficultyObject = {
    1: 'Fácil',
    2: 'Fácil',
    3: 'Fácil',
    4: 'Fácil',
    5: 'Médio',
    6: 'Médio',
    7: 'Médio',
    8: 'Difícil',
    9: 'Difícil',
    10: 'Difícil',
  };

  const statusMap = {
    A: 'Aprovado',
    P: 'Pendente',
    R: 'Reprovado',
  };

  return (
    <Grow in>
      <MaterialCard className={classes.root} {...others}>
        <CardActionArea
          onClick={handleOnClickCard}
          style={{ position: 'relative' }}
        >
          <div className={classes.tag}>
            {showStatus ? statusMap[status] : difficultyObject[difficulty]}
          </div>
          {logo === undefined ? (
            <div
              style={{
                display: 'flex',
                justifyContent: 'center',
                alignItems: 'center',
                minHeight: 200,
              }}
            >
              <CircularProgress color="secondary" />
            </div>
          ) : logo === null ? (
            <CardMedia
              component="img"
              height={200}
              image={getPlaceholder(recipe_id)}
              title="Default image"
            />
          ) : (
            <CardMedia
              height={200}
              component="img"
              image={logo}
              title="Main image"
            />
          )}
          <CardContent className={classes.content}>
            <Typography
              style={{ fontWeight: '500' }}
              variant="h5"
              component="h2"
            >
              {title}
            </Typography>
          </CardContent>
        </CardActionArea>

        <CardActions className={classes.actions}>
          <div
            style={{
              display: 'flex',
              alignItems: 'center',
              justifyContent: 'center',
            }}
          >
            <Clock />
            <Typography style={{ marginLeft: 8 }}>
              {preparation_time} min
            </Typography>
          </div>
          <div
            style={{
              display: 'flex',
              alignItems: 'center',
              justifyContent: 'center',
            }}
          >
            <Plate />
            <Typography style={{ marginLeft: 8 }}>
              {portions} porções
            </Typography>
          </div>
          <IconButton
            className={clsx(classes.expand, {
              [classes.expandOpen]: expanded,
            })}
            onClick={handleExpandClick}
            aria-expanded={expanded}
            aria-label="show more"
          >
            <ExpandMoreIcon />
          </IconButton>
        </CardActions>
        <Collapse in={expanded} timeout="auto" unmountOnExit>
          <CardContent>
            <div
              style={{
                display: 'flex',
                alignItems: 'center',
                justifyContent: 'space-between',
              }}
            >
              <Typography style={{ fontWeight: '500' }}>Descrição</Typography>
              <IconButton
                aria-label="add to favorites"
                color={isLiked ? 'secondary' : 'default'}
                onClick={() => handleHeartClick()}
              >
                <FavoriteIcon />
              </IconButton>
            </div>
            <Typography>{description}</Typography>
          </CardContent>
        </Collapse>
      </MaterialCard>
    </Grow>
  );
}

Card.propTypes = {
  data: PropTypes.object,
  onClickCard: PropTypes.func,
};

Card.defaultProps = {
  data: {},
  onClickCard: () => {},
};

export default Card;
