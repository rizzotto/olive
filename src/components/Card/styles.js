import { makeStyles } from '@material-ui/styles';

function getDifficultyColor(difficulty) {
  const difficultyColorMap = {
    1: '#52944F',
    2: '#52944F',
    3: '#52944F',
    4: '#52944F',
    5: '#FFCE00',
    6: '#FFCE00',
    7: '#FFCE00',
    8: '#FF6E61',
    9: '#FF6E61',
    10: '#FF6E61',
  };

  return difficultyColorMap[difficulty];
}

function getStatusColor(status) {
  const statusColorMap = {
    A: '#52944F',
    P: '#FFCE00',
    R: '#FF6E61',
  };

  return statusColorMap[status];
}

const useStyles = makeStyles(theme => ({
  root: {
    borderRadius: 20,
  },
  expand: {
    transform: 'rotate(0deg)',
    marginLeft: 'auto',
    transition: theme.transitions.create('transform', {
      duration: theme.transitions.duration.shortest,
    }),
  },
  expandOpen: {
    transform: 'rotate(180deg)',
  },
  content: {
    '&.MuiCardContent-root': {
      padding: '6px 16px',
    },
  },
  actions: {
    width: '100%',
    justifyContent: 'space-between',
    '&.MuiCardActions-root': {
      '&.MuiCardActions-spacing': {
        padding: '0 5px 0 16px',
      },
    },
  },
  tag: {
    position: 'absolute',
    textAlign: 'center',
    top: 10,
    left: 10,
    color: 'white',
    backgroundColor: props =>
      props.showStatus
        ? getStatusColor(props.status)
        : getDifficultyColor(props.difficulty),
    padding: '6px 10px',
    borderRadius: 21,
    letterSpacing: 0.9,
  },
}));

export default useStyles;
