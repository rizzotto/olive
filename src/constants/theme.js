import { createMuiTheme } from '@material-ui/core/styles';

const primaryGradient = gradient => {
  return `rgba(53, 119, 53, ${gradient})`;
};

const theme = createMuiTheme({
  palette: {
    primary: {
      main: primaryGradient(1),
      gradient80: primaryGradient(0.8),
      gradient60: primaryGradient(0.6),
      gradient40: primaryGradient(0.4),
      gradient20: primaryGradient(0.2),
      gradient10: primaryGradient(0.1),
    },
    base: {
      primary: '#FEF9EE',
      secondary: '#F6EDD9',
      tertiary: '#EFEEE5',
    },
    secondary: {
      main: '#52944F',
    },
    tertiary: {
      main: '#6FB26B',
    },
    quaternary: {
      main: '#8DD187',
    },
    grey: {
      main: '#F3F3F3',
      medium: '#DCDCDD',
      darker: '#8A8A8E',
    },
    contrastThreshold: 3,
    tonalOffset: 0.2,
  },
  typography: {
    fontFamily: 'Montserrat, Helvetica, Arial, sans-serif',
    color: '#212121',
  },
});

export default theme;
