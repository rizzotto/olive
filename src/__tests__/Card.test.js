import React from 'react';
import renderer from 'react-test-renderer';
import Card from '../components/Card';
import { ThemeProvider } from '@material-ui/core';
import theme from '../constants/theme';

test('renders a simple Card', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={theme}>
        <Card />
      </ThemeProvider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
