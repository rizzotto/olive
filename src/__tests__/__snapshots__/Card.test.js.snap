// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a simple Card 1`] = `
<div
  className="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
>
  <button
    className="MuiButtonBase-root MuiCardActionArea-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "position": "relative",
      }
    }
    tabIndex={0}
    type="button"
  >
    <div
      className="makeStyles-tag-6 makeStyles-tag-7"
    />
    <img
      className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
      src="recipePlaceholder.png"
      title="Contemplative Reptile"
    />
    <div
      className="MuiCardContent-root makeStyles-content-4"
    >
      <h2
        className="MuiTypography-root MuiTypography-h5"
        style={
          Object {
            "fontWeight": "500",
          }
        }
      />
    </div>
    <span
      className="MuiCardActionArea-focusHighlight"
    />
  </button>
  <div
    className="MuiCardActions-root makeStyles-actions-5 MuiCardActions-spacing"
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "center",
        }
      }
    >
      <svg>
        clock.svg
      </svg>
      <p
        className="MuiTypography-root MuiTypography-body1"
        style={
          Object {
            "marginLeft": 8,
          }
        }
      >
         Min
      </p>
    </div>
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "center",
        }
      }
    >
      <svg>
        plate.svg
      </svg>
      <p
        className="MuiTypography-root MuiTypography-body1"
        style={
          Object {
            "marginLeft": 8,
          }
        }
      >
         por√ßoes
      </p>
    </div>
    <button
      aria-expanded={false}
      aria-label="show more"
      className="MuiButtonBase-root MuiIconButton-root makeStyles-expand-2"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
